Add duration and unit kwargs to __init__ for all standard gates

alekkemeny opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Add duration and unit kwarg in __init__ for all classes in standard_gates


### Details and comments

alekkemeny committed:
Add duration and unit to kwargs for all standard_gates

alekkemeny committed:
Add release notes

mtreinish commented on releasenotes/notes/add-kwargs-to-standard-gates-d3ebf755e1d30efd.yaml:
```diff
@@ -0,0 +1,4 @@
+fixes:
+  - |
+    Added kwargs to standard gates for duration and unit.
```
Thanks for being through and adding a release note, but we actually don't need one for this PR. We already documented this fix in https://github.com/Qiskit/qiskit/blob/main/releasenotes/notes/singletons-83782de8bd062cbc.yaml#L118 which hasn't been included in a release yet. This PR is fixing the fix to actually be complete (as opposed to a subset of gate classes) :)

jakelishman commented on qiskit/circuit/library/standard_gates/global_phase.py:
```diff
@@ -34,13 +34,15 @@ class GlobalPhaseGate(Gate):
-    def __init__(self, phase: ParameterValueType, label: Optional[str] = None):
+    def __init__(
+        self, phase: ParameterValueType, label: Optional[str] = None, duration=None, unit=None
+    ):
```
Since we're defining new keyword arguments, we have the option to make them keyword-only (so you _can't_ pass them positionally like `XGate("my label", 1.0, "dt")` but have to say `XGate(duration=1.0, unit="dt")`).  We can put a `*` between the last allowed-to-be positional argument and the first kwarg, e.g.
```suggestion
    def __init__(
        self, phase: ParameterValueType, label: Optional[str] = None, *, duration=None, unit=None
    ):
```

In #11014 we're modifying all the singleton standard gate signatures to have that, so it'd be good to have it for these as well.

jakelishman reviewed (commented):
Thanks for this!  Could we do the same for the two standard instructions `Measure` and `Reset` as well?  (It doesn't make sense for `Barrier`, since that's a directive that's ignored by hardware.)

alekkemeny committed:
Make duration and unit keyword only

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into kwargs-fix

jakelishman committed:
Normalise all standard-library gate signatures

This corrects a couple of places that had `unit=None` instead of
`unit="dt", and makes sure that all gates in a given file have the right
signature.

jakelishman committed:
Fix bad default argument

jakelishman reviewed (approved):
Thanks for doing this, Hunter - I've fixed up the merge conflicts and just tweaked a couple of gates where there were typos (like passing `unit="dt"` on to `super().__init__` rather than `unit=unit`).
